include/os/freebsd/zfs/sys/zfs_context_os.h

```diff
@@ -89,4 +89,25 @@ extern int spa_import_rootpool(const char *name);
 #if BYTE_ORDER != BIG_ENDIAN
 #undef _BIG_ENDIAN
 #endif
+
+#define	zuio_segflg(U)		(U)->uio_segflg
```

Comment:
Let me make an alternate proposal.  How about something along these lines. 

```c
#define uio_segflg(uio)         (uio)->uio_segflg
#define uio_offset(uio)         (uio)->uio_loffset
#define uio_resid(uio)          (uio)->uio_resid
#define uio_iovcnt(uio)         (uio)->uio_iovcnt
#define uio_iovlen(uio, idx)    (uio)->uio_iov[(idx)].iov_len
#define uio_iovbase(uio, idx)   (uio)->uio_iov[(idx)].iov_base

static inline void
uio_advance(uio_t *uio, size_t size)
{
        uio->uio_resid -= size;
        uio->uio_loffset += size;
}

static inline offset_t
uio_index_at_offset(uio_t *uio, offset_t off, uint_t *vec_idx)
{
        while (*vec_idx < uio_iovcnt(uio) && *off >= uio_iovlen(uio, *vec_idx) {
                *off -= uio_iovlen(*vec_idx);
                vec_idx++;
        }

        return (off);
}
```

I don't believe any of these names conflict with existing functions and it'd be nice to stick with the `uio_` prefix if we can.  I renamed `zuio_update` to `uio_advance` which felt a little more descriptive, and I took a stab at reworking `zuio_nonemptyindex`.  Though I feel like this could all still be simpler with the right helper function.  I'm not sure zuio_iov` is needed either, why do we need a macro for this assignment?  If we do that's fine, but let's fine a better name.  Mind you none of this has been tested.  :)  What are your thoughts.

Reply:
I did not want to force you (upstream) into APPLE API, since they define
uio_offset
uio_resid
uio_iovcnt
uio_update (resolved by uio_advance)

But, those happen to be compatible. It is the APPLE uio_iovlen uio_iovbase and uio_iov that work completely differently, however, they have different name (uio_getiov()) so that would work as well.

`zuio_nonemptyindex()` I pasted it verbatim, as to not risk adding a bug to your code :) The OsX one I had to cook up.

zuio_iov() are all over module/icp:
```
 algs/modes/modes.c:         zuio_iov(uio, vec_idx, iov_base, iov_len);
```

So yes, I can do the zuio to uio rename I think.

Reply:
Oh uio_iov() is not an assignment, it's getting both len and base in one call.

Reply:
OK, then let's try the proposed rename if it's not going to cause any trouble on your side.  Then we can take it from there.

Regarding `zuio_nonemptyindex()` I appreciate the concern!  We could go with what you have, but this seems like as good opportunity as any to simplify that area of the code if we can.

For `uio_iov()` let's come up with a more descriptive name and add a comment explaining what it does.  :)

Reply:
Ah I see what you mean, we don't need uio_iov as we already have uio_iovlen and uio_iovbase - it was done for benefit of osx, since `uio_getiov(uio, &base, &len);`. If all calls were changed to separate calls, then osx has to call `uio_getiov()` twice, when it can fetch both in one go.